src/resource-metadata/lambda.js

```diff
@@ -0,0 +1,227 @@
+    const { functionResources, aliasResources, versionsToCollect } = await collectFunctionAndAliasResources(listOfFunctions)
+
+    console.info("Collecting function version details")
+    const functionVersionResources = await collectFunctionVersionResources(versionsToCollect)
+
+    const resources = [...functionResources, ...functionVersionResources, ...aliasResources]
+
+    resources.forEach(f =>
+        console.debug(`Resource: ${JSON.stringify(f)}`)
+    )
+
+    return resources
+}
+
+const collectFunctionAndAliasResources = async (listOfFunctions) => {
+    // This is important, because it avoids hitting the AWS API rate limits.
+    // The AWS API rate limit is 15req/s. rateLimit is typically specified to a lower value to leave some margin.
+    // The rateLimit is divided by 2, because the code it does two (GetFunction count's differently) calls per function.
+    const functionThrottle = pThrottle({ limit: rateLimit / 2, interval: 1000 })
+
+    // This is just an additional protection against doing to much in parallel, but typically the throttle should kick in first.
+    const limit = pLimit(10);
+
+    const results = (await traverse(listOfFunctions.Functions, (lambdaFunctionVersionLatest) => {
+        return limit(functionThrottle(async () => {
+            const functionName = lambdaFunctionVersionLatest.FunctionName
+            const lambdaFunction = await lambdaClient.send(new GetFunctionCommand({ FunctionName: functionName }))
+            const versions = await lambdaClient.send(new ListVersionsByFunctionCommand({ FunctionName: functionName }))
+            const aliases = await lambdaClient.send(new ListAliasesCommand({ FunctionName: functionName }))
+
+            const versionsToCollect = versions.Versions.filter((version, index) => {
+                return (index <= latestVersionsPerFunction) // Is either $LATEST or one of the latestVersionsPerFunction latest released versions // This relies on the fact that AWS returns the functions in latest -> oldest order
+                    || (aliases.Aliases.some(alias => version.Version === alias.FunctionVersion)) // has an alias
+            })
+
+            return {
+                functionResource: makeLambdaFunctionResource(lambdaFunction),
+                aliasResources: aliases.Aliases.map(alias => makeAliasResource(functionName, alias)),
+                versionsToCollect,
+            }
+        }))
+    }))
+
+    return {
+        functionResources: results.map(x => x.functionResource),
+        aliasResources: results.flatMap(x => x.aliasResources),
+        versionsToCollect: results.flatMap(x => x.versionsToCollect),
+    }
+}
+
+const collectFunctionVersionResources = async (versionsToCollect) => {
+    // The rateLimit is divided by 2, because the code it does two (ListEventSourceMappingsCommand, GetPolicyCommand) calls per function version.
+    const functionVersionThrottle = pThrottle({ limit: rateLimit / 2, interval: 1000 })
```

Comment:
This only works if this lambda is the only client of the lambda API, right? Otherwise, it will anyway fall back to SDK built-in retries.
Also, if a limit has been increased for the AWS account, the collection will slower than necessary. 
Are we actually winning anything compared to a simple sequential approach? Aren't the limits low enough to make any kind of parallelism useless?

Reply:
> This only works if this lambda is the only client of the lambda API, right? 

I set the default rateLimit to 10req/s and the AWS APIs limit is 15 req/s. The idea is that this leaves some headroom, so that other things can use the API as well. I made it configurable, so that the user can adjust it if needed.

> Otherwise, it will anyway fall back to SDK built-in retries.

Whatever is build into the SDK is insufficient. If I increase or remove the `rateLimit`, the lambda fails with `TooManyRequestsException`. I didn't dig into what is built-in and if it can be tuned.

> Are we actually winning anything compared to a simple sequential approach? Aren't the limits low enough to make any kind of parallelism useless?

We are, but not much. And I've tested locally, not on lambda. The execution time is ~32s with current parallelization and ~40s without it.

While relaying on retries would be better for squeezing the best throughput, it would also mean that we would often utilize 100% of available API quota, and I think this would be bad for any other program trying to use the APIs, so IMO having a rate limit that the user can control is a way to ensure that we're a "good neighbour".

Going with sequential and no throttling is a tempting option, as it seems to just work, although I'm not sure if it's a given or I'm just lucky :/ .

Reply:
I think the build-in retries are good enough for "single threaded" load, and they only fail when I parallelize the workload.